src/meta/src/rpc/leader_svc.rs

```diff
@@ -0,0 +1,343 @@
+use risingwave_pb::ddl_service::ddl_service_server::DdlServiceServer;
+use risingwave_pb::health::health_server::HealthServer;
+use risingwave_pb::hummock::hummock_manager_service_server::HummockManagerServiceServer;
+use risingwave_pb::meta::cluster_service_server::ClusterServiceServer;
+use risingwave_pb::meta::heartbeat_service_server::HeartbeatServiceServer;
+use risingwave_pb::meta::notification_service_server::NotificationServiceServer;
+use risingwave_pb::meta::scale_service_server::ScaleServiceServer;
+use risingwave_pb::meta::stream_manager_service_server::StreamManagerServiceServer;
+use risingwave_pb::meta::MetaLeaderInfo;
+use risingwave_pb::user::user_service_server::UserServiceServer;
+use tokio::sync::oneshot::Sender as OneSender;
+use tokio::sync::watch::Receiver as WatchReceiver;
+use tokio::task::JoinHandle;
+
+use super::intercept::MetricsMiddlewareLayer;
+use super::service::health_service::HealthServiceImpl;
+use super::service::notification_service::NotificationServiceImpl;
+use super::service::scale_service::ScaleServiceImpl;
+use super::DdlServiceImpl;
+use crate::backup_restore::BackupManager;
+use crate::barrier::{BarrierScheduler, GlobalBarrierManager};
+use crate::hummock::{CompactionScheduler, HummockManager};
+use crate::manager::{
+    CatalogManager, ClusterManager, FragmentManager, IdleManager, MetaOpts, MetaSrvEnv,
+};
+use crate::rpc::metrics::MetaMetrics;
+use crate::rpc::server::AddressInfo;
+use crate::rpc::service::backup_service::BackupServiceImpl;
+use crate::rpc::service::cluster_service::ClusterServiceImpl;
+use crate::rpc::service::heartbeat_service::HeartbeatServiceImpl;
+use crate::rpc::service::hummock_service::HummockServiceImpl;
+use crate::rpc::service::stream_service::StreamServiceImpl;
+use crate::rpc::service::user_service::UserServiceImpl;
+use crate::storage::MetaStore;
+use crate::stream::{GlobalStreamManager, SourceManager};
+use crate::{hummock, MetaResult};
+
+// simple wrapper containing election sync related objects
+pub struct ElectionCoordination {
+    pub election_handle: JoinHandle<()>,
+    pub election_shutdown: OneSender<()>,
+}
+
+/// Starts all services needed for the meta leader node
+/// Only call this function once, since initializing the services multiple times will result in an
+/// inconsistent state
+///
+/// ## Returns
+/// Returns an error if the service initialization failed
+pub async fn start_leader_srv<S: MetaStore>(
+    meta_store: Arc<S>,
+    address_info: AddressInfo,
+    max_heartbeat_interval: Duration,
+    opts: MetaOpts,
+    leader_rx: WatchReceiver<(MetaLeaderInfo, bool)>,
+    election_coordination: ElectionCoordination,
+    mut svc_shutdown_rx: WatchReceiver<()>,
+) -> MetaResult<()> {
+    tracing::info!("Defining leader services");
+    let prometheus_endpoint = opts.prometheus_endpoint.clone();
+    let env = MetaSrvEnv::<S>::new(opts, meta_store.clone(), leader_rx).await;
```

Comment:
If we only initialize `MetaSrvEnv` in leader service, it seems that updating leader info in env is not necessary. When follower becomes new leader, it will start leader service with new leader information. So the original leader should panic when committing txn in hummock using Illegal leader info in its env, which means some other node has become new leader. It's part of fencing mechanism.

Reply:
Okay. I think I get your point now. My suggestion is to keep it for now, since we do not have a fencing mechanism just yet. I added this task in the [fencing mechanism issue](https://github.com/risingwavelabs/risingwave/issues/6786). Are you fine with this approach?

Reply:
I think you can simply remove/revert the update logic for leader info in this PR. It won’t affect any consistency or correctness and also can expose bugs more easily in the unit tests if there's something wrong with the failover logic. It's not even a subtask, just some code clean. 😄

Reply:
Done. You are right: Undoing the change here and simply passing around a `MetaLeaderInfo` is a simple solution.